module/state_synchronization/indexer/indexer_core.go

```diff
@@ -0,0 +1,178 @@
+	headers   storage.Headers
+	events    storage.Events
+	log       zerolog.Logger
+}
+
+// New execution state indexer used to ingest block execution data and index it by height.
+// The passed RegisterIndex storage must be populated to include the first and last height otherwise the indexer
+// won't be initialized to ensure we have bootstrapped the storage first.
+func New(
+	log zerolog.Logger,
+	registers storage.RegisterIndex,
+	headers storage.Headers,
+	events storage.Events,
+) (*IndexerCore, error) {
+	return &IndexerCore{
+		registers: registers,
+		headers:   headers,
+		events:    events,
+		log:       log.With().Str("component", "execution_indexer").Logger(),
+	}, nil
+}
+
+// RegisterValues retrieves register values by the register IDs at the provided block height.
+// Even if the register wasn't indexed at the provided height, returns the highest height the register was indexed at.
+// Expected errors:
+// - storage.ErrNotFound if the register by the ID was never indexed
+// - ErrIndexBoundary if the height is out of indexed height boundary
+func (i *IndexerCore) RegisterValues(IDs flow.RegisterIDs, height uint64) ([]flow.RegisterValue, error) {
+	values := make([]flow.RegisterValue, len(IDs))
+
+	for j, id := range IDs {
+		value, err := i.registers.Get(id, height)
+		if err != nil {
+			return nil, err
+		}
+
+		values[j] = value
+	}
+
+	return values, nil
+}
+
+// IndexBlockData indexes all execution block data by height.
+// This method shouldn't be used concurrently.
+// Expected errors:
+// - ErrIndexValue if height was not incremented in sequence
+// - storage.ErrNotFound if the block for execution data was not found
+func (i *IndexerCore) IndexBlockData(_ context.Context, data *execution_data.BlockExecutionDataEntity) error {
+	fmt.Println("index block data", data.BlockID)
+	block, err := i.headers.ByBlockID(data.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not get the block by ID %s: %w", data.BlockID, err)
+	}
+
+	lg := i.log.With().
+		Hex("block_id", logging.ID(data.BlockID)).
+		Uint64("height", block.Height).
+		Logger()
+
+	lg.Debug().Msgf("indexing new block")
+
+	// the height we are indexing must be exactly one bigger or same as the latest height indexed from the storage
+	latest := i.registers.LatestHeight()
+	if block.Height != latest+1 && block.Height != latest {
+		return fmt.Errorf("must store registers with the next height %d, but got %d", latest+1, block.Height)
+	}
+	// allow rerunning the indexer for same height since we are fetching height from register storage, but there are other storages
+	// for indexing resources which might fail to update the values, so this enables rerunning and reindexing those resources
+	if block.Height == latest {
+		lg.Warn().Msg("reindexing block data")
+	}
+
+	// concurrently process indexing of block data
+	g := errgroup.Group{}
+
+	updates := make([]*ledger.TrieUpdate, 0)
+	events := make([]flow.Event, 0)
+
+	for _, chunk := range data.ChunkExecutionDatas {
```

Comment:
can we move this to the `g.Go` so that it doesn't block indexing the registers

Reply:
you mean block events? it will have to block registers, but for events I agree, good point